Sources/PackageGraph/RegistryPackageContainer.swift

```diff
@@ -0,0 +1,207 @@
+ */
+
+import Basics
+import PackageLoading
+import PackageModel
+import PackageRegistry
+import SourceControl
+
+import TSCBasic
+import TSCUtility
+
+import Dispatch
+
+public final class RegistryPackageContainer {
+    public var identifier: PackageReference
+
+    let mirrors: DependencyMirrors
+
+    /// The manifest loader.
+    let manifestLoader: ManifestLoaderProtocol
+
+    /// The tools version loader.
+    let toolsVersionLoader: ToolsVersionLoaderProtocol
+
+    /// The current tools version in use.
+    let currentToolsVersion: ToolsVersion
+
+    public let registryManager: RegistryManager
+
+    private(set) var reversedVersionsCache: [Version] = []
+    private var manifestsCache = ThreadSafeKeyValueStore<Version, Manifest>()
+    private var toolsVersionsCache = ThreadSafeKeyValueStore<Version, ToolsVersion>()
+    private var validToolsVersionsCache = ThreadSafeKeyValueStore<Version, Bool>()
+
+    private init(
+        _ identifier: PackageReference,
+        mirrors: DependencyMirrors,
+        registryManager: RegistryManager,
+        manifestLoader: ManifestLoaderProtocol,
+        toolsVersionLoader: ToolsVersionLoaderProtocol,
+        currentToolsVersion: ToolsVersion
+    ) {
+        self.identifier = identifier
+        self.mirrors = mirrors
+        self.registryManager = registryManager
+        self.manifestLoader = manifestLoader
+        self.toolsVersionLoader = toolsVersionLoader
+        self.currentToolsVersion = currentToolsVersion
+    }
+
+    static func create(
+        for identifier: PackageReference,
+        mirrors: DependencyMirrors,
+        registryManager: RegistryManager,
+        manifestLoader: ManifestLoaderProtocol,
+        toolsVersionLoader: ToolsVersionLoaderProtocol,
+        currentToolsVersion: ToolsVersion,
+        completion: (Result<RegistryPackageContainer, Error>) -> Void
+    ) {
+        let container = RegistryPackageContainer(identifier,
+                                                 mirrors: mirrors,
+                                                 registryManager: registryManager,
+                                                 manifestLoader: manifestLoader,
+                                                 toolsVersionLoader: toolsVersionLoader,
+                                                 currentToolsVersion: currentToolsVersion)
+
+        do {
+            let versionsWithDuplicates = try tsc_await { registryManager.fetchVersions(of: identifier, completion: $0) }
+            container.reversedVersionsCache = Set(versionsWithDuplicates).sorted(by: >)
+            completion(.success(container))
+        } catch {
+            completion(.failure(error))
+        }
+    }
+
+    private func manifest(for version: Version) throws -> Manifest {
+        try manifestsCache.memoize(version) {
+            try tsc_await { registryManager.fetchManifest(for: version, of: identifier, using: manifestLoader, completion: $0) }
```

Comment:
we should probably turn manifest(for version: Version) into an async function, it makes sense in all cases really